libraries/Microsoft.Bot.Builder.Runtime/CoreBotAdapter.cs

```diff
@@ -0,0 +1,49 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using Microsoft.Bot.Builder;
+using Microsoft.Bot.Builder.Integration.AspNet.Core;
+using Microsoft.Bot.Builder.Runtime.Builders.Middleware;
+using Microsoft.Bot.Builder.Runtime.Settings;
+using Microsoft.Bot.Connector.Authentication;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Bot.Builder.Runtime
+{
+    public class CoreBotAdapter : BotFrameworkHttpAdapter
+    {
+        public CoreBotAdapter(
+            IServiceProvider services,
+            IConfiguration configuration,
+            IOptions<CoreBotAdapterOptions> options)
+            : base(
+                services.GetService<ICredentialProvider>(),
+                services.GetService<AuthenticationConfiguration>(),
+                services.GetService<IChannelProvider>(),
+                logger: services.GetService<ILogger<BotFrameworkHttpAdapter>>())
+        {
+            var conversationState = services.GetService<ConversationState>();
+            var userState = services.GetService<UserState>();
+
+            this.UseStorage(services.GetService<IStorage>());
+            this.UseBotState(userState, conversationState);
+            this.Use(new RegisterClassMiddleware<IConfiguration>(configuration));
+
+            foreach (IMiddlewareBuilder middleware in options.Value.Middleware)
```

Comment:
Yes, there is! The reason for this is that, going forward, one of the scenarios we need to support is multiple adapter configurations, each of which may have their own middleware pipeline configurations. In theory, you could have two separate instances of the same type of middleware, each with their own different configuration.

If we follow standard DI, then as things are right now, there is no way to really differentiate from IServiceProvider which instance is mapped to each adapter. The clean option would be to the use the [Options pattern](https://docs.microsoft.com/en-us/aspnet/core/fundamentals/configuration/options?view=aspnetcore-5.0), but this would require I think changes in other layers in the SDK to more easily support.

For now, we opted that the right pattern to use is let each adapter manage its own middleware instances, hence why we've taken this approach. In the future, I would like to evaluate if we can just make use of the Options pattern instead.